src/main/java/io/github/thebusybiscuit/slimefun4/implementation/operations/CraftingOperation.java

```diff
@@ -0,0 +1,58 @@
+package io.github.thebusybiscuit.slimefun4.implementation.operations;
+
+import javax.annotation.Nonnull;
+
+import org.apache.commons.lang.Validate;
+import org.bukkit.inventory.ItemStack;
+
+import io.github.thebusybiscuit.slimefun4.core.machines.MachineOperation;
+import me.mrCookieSlime.Slimefun.Objects.SlimefunItem.abstractItems.MachineRecipe;
+
+public class CraftingOperation implements MachineOperation {
+
+    private final ItemStack[] ingredients;
+    private final ItemStack[] results;
+
+    private final int totalTicks;
+    private int currentTicks = 0;
+
+    public CraftingOperation(@Nonnull MachineRecipe recipe) {
+        this(recipe.getInput(), recipe.getOutput(), recipe.getTicks());
+    }
+
+    public CraftingOperation(@Nonnull ItemStack[] ingredients, @Nonnull ItemStack[] results, int totalTicks) {
+        Validate.notEmpty(ingredients, "The Ingredients array cannot be empty or null");
+        Validate.notEmpty(results, "The results array cannot be empty or null");
+        Validate.isTrue(totalTicks > 0, "The amount of total ticks must be a positive integer");
+
+        this.ingredients = ingredients;
+        this.results = results;
+        this.totalTicks = totalTicks;
+    }
+
+    @Override
+    public void addProgress(int num) {
+        currentTicks += num;
```

Comment:
Maybe a check to make sure its a positive number?
Or start using a uint I suppose

Unless you want to support removing progress from machines, that is